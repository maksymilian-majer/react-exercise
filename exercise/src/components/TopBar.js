import React, { Component, Fragment } from "react";
import { Link } from "react-router-dom";
import { connect } from "react-redux";
import PropTypes from "prop-types";
import compose from "recompose/compose";
import { withRouter } from "react-router-dom";

import { withStyles } from "@material-ui/core/styles";
import AppBar from "@material-ui/core/AppBar";
import Toolbar from "@material-ui/core/Toolbar";
import Typography from "@material-ui/core/Typography";
import Button from "@material-ui/core/Button";
import IconButton from "@material-ui/core/IconButton";
import ArrowBack from "@material-ui/icons/ArrowBackIos";
import Avatar from "@material-ui/core/Avatar";

import profileActions from "../actions/profile";

const styles = {
  root: {
    flexGrow: 1
  },
  flex: {
    flex: 1
  },
  backButton: {
    marginLeft: -12,
    marginRight: 20
  },
  loginButton: {
    marginRight: 10
  },
  avatar: {
    marginRight: 10
  },
  title: {
    color: "#fff",
    textDecoration: "none"
  },
  logo: {
    marginRight: 10,
    maxHeight: 40,
    flex: 1
  }
};

class TopBar extends Component {
  renderLoggedOut() {
    return (
      <Fragment>
        <Button
          variant="contained"
          onClick={() => this.props.logIn("MM")}
          className={this.props.classes.loginButton}
        >
          Login
        </Button>
        <Button variant="contained" color="secondary">
          Signup
        </Button>
      </Fragment>
    );
  }

  renderLoggedIn() {
    return (
      <Fragment>
        <Avatar className={this.props.classes.avatar}>
          <Typography variant="body1">{this.props.initials}</Typography>
        </Avatar>
        <Button variant="contained" color="secondary" onClick={this.props.logOut}>
          Logout
        </Button>
      </Fragment>
    );
  }

  render() {
    const classes = this.props.classes;
    const backButton =
      this.props.history.location.pathname !== "/" ? (
        <IconButton
          className={classes.backButton}
          color="inherit"
          aria-label="Back"
          onClick={this.props.history.goBack}
        >
          <ArrowBack />
        </IconButton>
      ) : null;
    return (
      <div className={classes.root}>
        <AppBar position="static">
          <Toolbar>
            {backButton}
            <Typography variant="title" color="inherit" className={classes.flex}>
              <Link to="/" className={classes.title}>
                <img alt={"logo"} className={classes.logo} src="/favicon-196x196.png" />
                Modus Create
              </Link>
            </Typography>
            {this.props.loggedIn ? this.renderLoggedIn() : this.renderLoggedOut()}
          </Toolbar>
        </AppBar>
      </div>
    );
  }
}

TopBar.propTypes = {
  loggedIn: PropTypes.bool.isRequired,
  initials: PropTypes.string
};

const mapStateToProps = state => ({
  loggedIn: state.profile.loggedIn,
  initials: state.profile.initials
});

const mapDispatchToProps = {
  logIn: profileActions.logIn,
  logOut: profileActions.logOut
};

export default compose(
  withStyles(styles),
  connect(
    mapStateToProps,
    mapDispatchToProps
  )
)(withRouter(TopBar));
