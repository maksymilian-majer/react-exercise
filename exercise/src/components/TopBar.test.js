import React from "react";
import { shallow, mount } from "enzyme";

import { TopBar } from "./TopBar";

const setup = propOverrides => {
  const defaultProps = {
    loggedIn: false,
    history: {
      location: {
        pathname: "/"
      }
    },
    classes: {
      root: "root",
      loginButton: "login"
    }
  };
  const props = Object.assign(defaultProps, propOverrides);
  const component = shallow(<TopBar {...props} />);

  return { props, component, mount };
};

describe("(Component) Top Bar", () => {
  it("should render correctly when logged out", () => {
    const { component } = setup();
    expect(component).toMatchSnapshot();
  });

  it("should render correctly when logged in", () => {
    const { component } = setup({
      loggedIn: true,
      initials: "AB"
    });
    expect(component).toMatchSnapshot();
  });

  describe("login button", () => {
    it("should call log in on click", () => {
      const loginMock = jest.fn();
      const { component } = setup({
        logIn: loginMock
      });
      const loginButton = component.find(".login");
      loginButton.simulate("click");

      expect(loginMock).toBeCalledWith("MM");
    });
  });
});
