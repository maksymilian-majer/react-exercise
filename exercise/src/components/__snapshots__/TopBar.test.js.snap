// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`(Component) Top Bar should render correctly when logged in 1`] = `
<div
  className="root"
>
  <WithStyles(AppBar)
    position="static"
  >
    <WithStyles(Toolbar)>
      <WithStyles(Typography)
        color="inherit"
        variant="title"
      >
        <Link
          replace={false}
          to="/"
        >
          <img
            alt="logo"
            src="/favicon-196x196.png"
          />
          Modus Create
        </Link>
      </WithStyles(Typography)>
      <WithStyles(Avatar)>
        <WithStyles(Typography)
          variant="body1"
        >
          AB
        </WithStyles(Typography)>
      </WithStyles(Avatar)>
      <WithStyles(Button)
        color="secondary"
        variant="contained"
      >
        Logout
      </WithStyles(Button)>
    </WithStyles(Toolbar)>
  </WithStyles(AppBar)>
</div>
`;

exports[`(Component) Top Bar should render correctly when logged out 1`] = `
<div
  className="root"
>
  <WithStyles(AppBar)
    position="static"
  >
    <WithStyles(Toolbar)>
      <WithStyles(Typography)
        color="inherit"
        variant="title"
      >
        <Link
          replace={false}
          to="/"
        >
          <img
            alt="logo"
            src="/favicon-196x196.png"
          />
          Modus Create
        </Link>
      </WithStyles(Typography)>
      <WithStyles(Button)
        className="login"
        onClick={[Function]}
        variant="contained"
      >
        Login
      </WithStyles(Button)>
      <WithStyles(Button)
        color="secondary"
        variant="contained"
      >
        Signup
      </WithStyles(Button)>
    </WithStyles(Toolbar)>
  </WithStyles(AppBar)>
</div>
`;
